starter/src/index.js

```diff
@@ -0,0 +1,109 @@
+import Apify from 'apify';
+import { inspect } from 'util';
+import { convertInputToActorConfigs } from './lib/configs.js';
+import { waitForRunToFinishAndPushData, startRun } from './lib/startRunAndPool.js';
+
+const { log } = Apify.utils;
+const env = Apify.getEnv();
+
+Apify.main(async () => {
+    /** @type {import('../../common/types').ActorInputData} */
+    // @ts-ignore It's not null
+    const input = await Apify.getInput();
+    log.debug('Provided inputs:');
+    log.debug(inspect(input));
+
+    const { maxConcurrentDomainsChecked, urlsToCheck } = input;
+
+    // Log the input
+    log.info('Input provided:');
+    log.debug(inspect(input, false, 4));
+
+    /** @type {import('./types').FrontendActorState} */
+    // @ts-expect-error It's an object
+    const state = await Apify.getValue('STATE') ?? {
+        runConfigurations: [],
+        totalUrls: urlsToCheck.length,
+        checkerFinished: false,
+    };
+
+    Apify.events.on('migrating', async () => {
+        await Apify.setValue('STATE', state);
+    });
+
+    Apify.events.on('persistState', async () => {
+        await Apify.setValue('STATE', state);
+    });
+
+    setInterval(async () => {
+        await Apify.setValue('STATE', state);
+
+        log.debug('Internal state:');
+        log.debug(inspect(state, false, 3));
+    }, 10_000);
+
+    // If we haven't initialized the state yet, do it now
+    if (state.runConfigurations.length === 0 && !state.checkerFinished) {
+        state.runConfigurations = convertInputToActorConfigs(input);
+    }
+
+    // Sort state based on started runs
+    state.runConfigurations = state.runConfigurations.sort((_, b) => Number(Boolean(b.runId)));
+    await Apify.setValue('STATE', state);
+
+    log.info(`Preparing to process ${state.totalUrls} URLs...\n`);
+
+    /** @type {import('apify').RequestOptions[]} */
+    const sources = state.runConfigurations.map((actorInput, index) => ({
+        url: 'https://localhost',
+        uniqueKey: index.toString(),
+        userData: { actorInput },
+    }));
+
+    const requestList = await Apify.openRequestList(null, sources);
+
+    const runner = new Apify.BasicCrawler({
+        maxConcurrency: maxConcurrentDomainsChecked,
+        requestList,
+        handleRequestFunction: async ({ request }) => {
+            const { userData } = request;
+            /** @type {{ actorInput: import('../../common/types').PreparedActorConfig }} */
+            // @ts-expect-error JS-style casting
+            const { actorInput } = userData;
```

Comment:
I think this implementation would bug out after migration. If you use RequestList, the userData are not persisted (unlike for queue where everything is stored on Apify platform backend) so after migration, it would start a fresh run. (I'm not 100% sure if this was not adjusted lately in the SDK so feel free to check the code or consult the team :) ). So I recommend saving the mapping of request -> runId into the STATE as in this article: https://www.notion.so/apify/How-to-correctly-call-actors-from-an-actor-7b0553a240e84830ac7e627e7320001d

Reply:
This is where these lines are related: 

https://github.com/metalwarrior665/actor-website-checker/blob/92f0247a0e7be3847f9871ae6c53d3fd0c787b8a/starter/src/index.js#L85-L89

Since the list is built from the state, and the state has the runId set (thus already handling the request -> runId in state), it is saved on the platform and should revive easily (I'd need a long test in order to check this tho, maybe if you have some test data we can validate this?). That's also why the TODO is there, since right now I made it save the state on every actor start, so we have the run id, but I didn't know if that was needed. I guess because it's saved on migrating, it isn't needed?

Reply:
Yes, not needed to store it since it is either in memory or just gets saved by 'persistState' (also handles `aborting` event). It is just not easy for me to see how all the JS references lead to the state but I see what you mean :) You can actually test it by doing the graceful abort (fires off `persistState`) and then resurrect

Reply:
It worked! 🎉 I'll DM you a link that shows it in action